src/day23.scala

```diff
@@ -0,0 +1,151 @@
+// using scala 3.1.0
+
+package day23
+
+import scala.util.Using
+import scala.io.Source
+import Position.*
+import scala.annotation.tailrec
+
+@main def part1(): Unit =
+  val answer = part1(readInput())
+  println(s"The answer is: $answer")
+
+@main def part2(): Unit =
+  val answer = part2(readInput())
+  println(s"The answer is:\n$answer")
+
+def readInput(): String =
+  Using.resource(Source.fromFile("input/day23"))(_.mkString)
+
+type Energy = Int
+
+enum Amphipod:
+  case A, B, C, D
+
+  def energy: Energy = this match
+    case A => 1
+    case B => 10
+    case C => 100
+    case D => 1000
+
+  def destinationX: Int = this match
+    case A => 3
+    case B => 5
+    case C => 7
+    case D => 9
+
+  def nextType: Amphipod = this match
+    case A => B
+    case B => C
+    case C => D
+    case D => assert(false)
+
+
+case class Position(val x: Int, val y: Int)
+
+case class Situation(amphipodPositions: Map[Amphipod, Set[Position]], roomSize: Int):
+  val positionToAmphipod: Map[Position, Amphipod] = 
+    amphipodPositions.flatMap((amphipod, positions) => positions.map(position => position -> amphipod))
+
+  def move(amphipod: Amphipod): Set[(Situation, Energy)] =
+    for
+      from <- amphipodPositions(amphipod)
+      stop <- nextStops(amphipod, from)
+      path = getPath(from, stop)
+      if path.forall(isEmpty)
+    yield
+      val newPositions = amphipodPositions(amphipod) - from + stop
+      val energy = path.size * amphipod.energy
+      (copy(amphipodPositions + (amphipod -> newPositions)), energy)
+  
+  def isFinal =
+    amphipodPositions.forall { (amphipod, positions) =>
+      positions.forall(p => p.x == amphipod.destinationX)
+    }
+
+  private def getDestination(amphipod: Amphipod): Option[Position] =
+    val x = amphipod.destinationX
+    @tailrec
+    def tailApply(y: Int): Option[Position] =
+      positionToAmphipod.get(Position(x, y)) match
+        case Some(other) if other == amphipod => if y == 2 then None else tailApply(y - 1)
+        case Some(_) => None
+        case None => Some(Position(x, y))
+    tailApply(roomSize + 1)
+  
+  private def nextStops(amphipod: Amphipod, from: Position): Seq[Position] =
+    val destination = getDestination(amphipod)
+    from match
+      case Position(x, y) if x == amphipod.destinationX =>
+        destination match
+          case Some(destination) => Seq.empty
+          case None => hallwayStops
+      case Position(_, 1) => destination.toSeq
+      case _ => hallwayStops ++ destination
```

Comment:
Easy win here:
```suggestion
      case _ => hallwayStops
```
Any solution that involves going from one room directly to the destination is subsumed by first making a stop in the hallway between the rooms, then moving into the destination. So by exploring both going in the hallway *and* directly go to the destination, you're duplicating work.